kube/src/api/streaming.rs

```diff
@@ -0,0 +1,274 @@
+// Code to support streaming resources (attach/exec/portforward)
+use std::{
+    future::Future,
+    pin::Pin,
+    sync::Arc,
+    sync::Mutex,
+    task::{Context, Poll, Waker},
+};
+
+use k8s_openapi::apimachinery::pkg::apis::meta::v1::Status;
+
+use async_tungstenite::{
+    tokio::ConnectStream,
+    tungstenite::{self as ws, Message},
+    WebSocketStream,
+};
+use futures::{future::Either, SinkExt, Stream, StreamExt, TryStreamExt};
+use futures_util::future::select;
+use tokio::io::{AsyncRead, AsyncWrite, AsyncWriteExt, DuplexStream};
+use tokio_util::codec;
+
+// Internal state of an attached process
+struct AttachedProcessState {
+    waker: Option<Waker>,
+    finished: bool,
+    status: Option<Status>,
+    stdin_writer: Option<DuplexStream>,
+    stdout_reader: Option<DuplexStream>,
+    stderr_reader: Option<DuplexStream>,
+}
+
+/// Represents an attached process in a container for `attach` and `exec`.
+///
+/// Provides access to stdin/stdout/stderr if attached.
+/// Resolves when the connection terminates with an optional [`Status`].
+///
+/// [`Status`]: k8s_openapi::apimachinery::pkg::apis::meta::v1::Status
+pub struct AttachedProcess {
```

Comment:
While stdout/stderr bools are fine for now, I do feel that it would be useful to have access to these stream objects so that people could send input streams through something else (e.g. and make something like teleport's ssh web app).

Reply:
I'm pretty sure it's already supported, but maybe I'm misunderstanding. Users have access to the pipes, so writing to `stdin` writer does send the data to the attached container.
Can you elaborate?

Reply:
Sorry, I misunderstood this part of the code completely. I saw all the names for stdin and stdout as methods on `AttachedProcess` and thought they were hardcoded to `io::stdout` and the likes. Your explanation below made this more clear. I think I expected them to be more generically named, but it's ultimately propagated from a process call deep in kubernetes so it's probably appropriate.

As long as people are able to pipe arbitrary input streams to the writer returned by `stdin` then this is great!